metrics/prometheus/factory_test.go

```diff
@@ -0,0 +1,168 @@
+	assert.EqualValues(t, 2, m1.GetGauge().GetValue(), "%+v", m1)
+
+	m2 := findMetric(t, snapshot, "bender:rodriguez", map[string]string{"a": "b", "x": "z"})
+	assert.EqualValues(t, 4, m2.GetGauge().GetValue(), "%+v", m2)
+}
+
+func TestTimer(t *testing.T) {
+	registry := prometheus.NewPedanticRegistry()
+	f1 := New(registry, nil)
+	f2 := f1.Namespace("bender", map[string]string{"a": "b"})
+	f3 := f2.Namespace("", map[string]string{"a": "b"}) // essentially same as f2
+	t1 := f2.Timer("rodriguez", map[string]string{"x": "y"})
+	t2 := f2.Timer("rodriguez", map[string]string{"x": "z"})
+	t3 := f3.Timer("rodriguez", map[string]string{"x": "z"}) // same as g2
+	t1.Record(1 * time.Second)
+	t1.Record(2 * time.Second)
+	t2.Record(3 * time.Second)
+	t3.Record(4 * time.Second)
+
+	snapshot, err := registry.Gather()
+	require.NoError(t, err)
+
+	m1 := findMetric(t, snapshot, "bender:rodriguez", map[string]string{"a": "b", "x": "y"})
+	assert.EqualValues(t, 2, m1.GetHistogram().GetSampleCount(), "%+v", m1)
+	assert.EqualValues(t, 3, m1.GetHistogram().GetSampleSum(), "%+v", m1)
+	for _, bucket := range m1.GetHistogram().GetBucket() {
+		if bucket.GetUpperBound() < 1 {
+			assert.EqualValues(t, 0, bucket.GetCumulativeCount())
+		} else if bucket.GetUpperBound() < 2 {
+			assert.EqualValues(t, 1, bucket.GetCumulativeCount())
+		} else {
+			assert.EqualValues(t, 2, bucket.GetCumulativeCount())
+		}
+	}
+
+	m2 := findMetric(t, snapshot, "bender:rodriguez", map[string]string{"a": "b", "x": "z"})
+	assert.EqualValues(t, 2, m2.GetHistogram().GetSampleCount(), "%+v", m2)
+	assert.EqualValues(t, 7, m2.GetHistogram().GetSampleSum(), "%+v", m2)
+	for _, bucket := range m2.GetHistogram().GetBucket() {
+		if bucket.GetUpperBound() < 3 {
+			assert.EqualValues(t, 0, bucket.GetCumulativeCount())
+		} else if bucket.GetUpperBound() < 4 {
+			assert.EqualValues(t, 1, bucket.GetCumulativeCount())
+		} else {
+			assert.EqualValues(t, 2, bucket.GetCumulativeCount())
+		}
+	}
+}
+
+func TestTimerCustomBuckets(t *testing.T) {
+	registry := prometheus.NewPedanticRegistry()
+	f1 := New(registry, []float64{1.5})
+	// dot and dash in the metric name will be replaced with underscore
+	t1 := f1.Timer("bender.bending-rodriguez", map[string]string{"x": "y"})
+	t1.Record(1 * time.Second)
+	t1.Record(2 * time.Second)
+
+	snapshot, err := registry.Gather()
+	require.NoError(t, err)
+
+	m1 := findMetric(t, snapshot, "bender_bending_rodriguez", map[string]string{"x": "y"})
+	assert.EqualValues(t, 2, m1.GetHistogram().GetSampleCount(), "%+v", m1)
+	assert.EqualValues(t, 3, m1.GetHistogram().GetSampleSum(), "%+v", m1)
+	assert.Len(t, m1.GetHistogram().GetBucket(), 1)
+}
+
+func findMetric(t *testing.T, snapshot []*promModel.MetricFamily, name string, tags map[string]string) *promModel.Metric {
+	for _, mf := range snapshot {
+		if mf.GetName() == name {
+			for _, m := range mf.GetMetric() {
+				if len(m.GetLabel()) != len(tags) {
```

Comment:
cant a prometheus metric with the same name have different labels? I guess this is a testing function and I shouldn't read too much into it?

Reply:
the function findMetric only returns exact match. Also, as a general point, the whole reason for this PR is that once you define a metric with a certain set of labels (tags), Prometheus will not allow another one with the same name but different set of labels.